orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,272 @@
+		bf.Logger.Debugf("Shuffle Disabled.")
+		return false
+	}
+
+	shuffle := time.Now().After(bf.LastShuffledAt.Add(bf.ShuffleTimeout))
+	bf.Logger.Debugf("Shuffle Enabled. Time to shuffle %v", shuffle)
+	return shuffle
+}
+
+func (bf *BlockFetcher) isEndpointByzantine(ec EndpointCriteria) bool {
+	// probe the other endpoints
+
+	// choose (MaxByzantineFaults + 1) endpoints other than currentEndpoint
+	//currIdx, err := bf.getIdxForEndpoint(bf.currentEndpoint.Endpoint)
+	currIdx, _ := bf.getIdxForEndpoint(ec.Endpoint)
+	// TODO: handle above error case
+
+	endpointsToProbe := []EndpointCriteria{}
+	for i := int64(currIdx + 1); i <= (bf.MaxByzantineFaults + int64(currIdx)); i++ {
+		// add the next (bf.MaxByzantineFaults + 1)
+		endpointsToProbe = append(endpointsToProbe, bf.Endpoints[i%(int64(len(bf.Endpoints)))])
+	}
+
+	// add a Sync.WaitGroup and probe using go routines
+	waitGrp := sync.WaitGroup{}
+	// recieve the blocks using channels to a slice
+
+	// verify the slice and decide whether current node had byzantine behavior
+	type bps struct {
+		block *common.Block
+		ec    EndpointCriteria
+	}
+
+	bpsInfo := make(chan *bps, len(endpointsToProbe))
+
+	// start probing
+	for _, ec := range endpointsToProbe {
+		// get a puller
+		waitGrp.Add(1)
+		go func(ec EndpointCriteria) {
+			defer waitGrp.Done()
+
+			config := bf.Config
+			config.Endpoints = []EndpointCriteria{ec}
+			puller := bf.BlockSourceFactory(config)
+			defer puller.Close()
+
+			// set endpoints
+			puller.UpdateEndpoints([]EndpointCriteria{ec})
+
+			// pull block
+			// TODO: Later pull attestations at this point
+			block := puller.PullBlock(bf.currentSeq)
+			bpsInfo <- &bps{block, ec}
+		}(ec)
+
+	}
+	waitGrp.Wait()
+	//close channel
+	close(bpsInfo)
+
+	for b := range bpsInfo {
+		// verify the blocks or attestations returned here and compare with the
+		// previously pulled block
+		fmt.Println(b.ec.Endpoint)
+	}
+
+	return false
+}
+
+func (bf *BlockFetcher) isBlockWitheld() bool {
+	// this should keep the current endpoint as reference and try to detect if block witholding
+	// takes place by probing (MaxByzantineFaults + 1) endpoints
+	if bf.isEndpointByzantine(bf.currentEndpoint) {
+		bf.suspects.insert(bf.currentEndpoint.Endpoint)
+		return true
+	}
+
+	return false
+}
+
+func (bf *BlockFetcher) isEndpointSuspected() bool {
+	return bf.suspects.has(bf.currentEndpoint.Endpoint)
+}
+
+func (bf *BlockFetcher) PullBlock(seq uint64) *common.Block {
+	retriesLeft := bf.MaxPullBlockRetries
+
+	// TODO: Should we persist the suspect set across PullBlock ?
+	// For now let us keep it unique for every PullBlock request
+	bf.suspects = suspectSet{}
+
+	for {
+
+		if retriesLeft == 0 && bf.MaxPullBlockRetries > 0 {
+			bf.Logger.Errorf("Failed pulling block [%d]: retry count exhausted(%d)", seq, bf.MaxPullBlockRetries)
+			return nil
+		}
+
+		bf.setSomeEndpoint()
+
+		block := bf.pullBlockFromSource(seq)
+		switch block {
```

Comment:
why do you use a switch and not if? you only have two cases.

Reply:
can use if here.